lms/djangoapps/certificates/api.py

```diff
@@ -463,49 +464,67 @@ def get_active_web_certificate(course, is_preview_mode=None):
+
+    if org_id and mode:  # get template by org, mode, and key
+        org_mode_and_key_templates = active_templates.filter(
             organization_id=org_id,
-            course_key=course_key,
             mode=mode,
-            is_active=True
+            course_key=course_key
         )
-    # if don't template find by org and mode
-    if not template and org_id and mode:
-        template = CertificateTemplate.objects.filter(
+        template = get_language_specific_template_or_default(language, org_mode_and_key_templates)
+
+    # since no template matched that course_key, only consider templates with empty course_key
+    empty_course_key_templates = active_templates.filter(course_key=CourseKeyField.Empty)
+    if not template and org_id and mode:  # get template by org and mode
+        org_and_mode_templates = empty_course_key_templates.filter(
             organization_id=org_id,
-            course_key=CourseKeyField.Empty,
-            mode=mode,
-            is_active=True
+            mode=mode
         )
-    # if don't template find by only org
-    if not template and org_id:
-        template = CertificateTemplate.objects.filter(
+        template = get_language_specific_template_or_default(language, org_and_mode_templates)
+    if not template and org_id:  # get template by only org
+        org_templates = empty_course_key_templates.filter(
             organization_id=org_id,
-            course_key=CourseKeyField.Empty,
-            mode=None,
-            is_active=True
+            mode=None
         )
-    # if we still don't template find by only course mode
-    if not template and mode:
-        template = CertificateTemplate.objects.filter(
+        template = get_language_specific_template_or_default(language, org_templates)
+    if not template and mode:  # get template by only mode
+        mode_templates = empty_course_key_templates.filter(
             organization_id=None,
-            course_key=CourseKeyField.Empty,
-            mode=mode,
-            is_active=True
+            mode=mode
         )
+        template = get_language_specific_template_or_default(language, mode_templates)
+    #return template[0].template if template else None
+    return template.template if template else None
+
 
-    return template[0].template if template else None
+def get_language_specific_template_or_default(language, templates):
+    """
+    Returns templates that match passed in language.
+    Returns default templates If no language matches, or language passed is None
+    """
+    two_letter_language = _get_two_letter_language_code(language)
+    language_or_default_templates = templates.filter(Q(language=two_letter_language) | Q(language=None))
+    language_or_default_templates.order_by('language')
+    # order_by sorts nulls to the top, so last template is a language_specific if there is one
+    return language_or_default_templates.last() if language_or_default_templates else None
```

Comment:
@bderusha @amangano-edx This is how I had previously rewritten the language template function, I believe it now only hits the DB only once.

Reply:
Offline we discussed filtering in python.